api/src/test/java/org/openmrs/api/EncounterServiceTest.java

```diff
@@ -98,7 +97,235 @@
+		ConceptService cs = Context.getConceptService();
+		EncounterService es = Context.getEncounterService();
+		ObsService os = Context.getObsService();
+
+		//create an obs
+		Obs o = createObs(cs.getConcept(3),"original obs value text");
+		//create a second obs
+		Obs o2 = createObs(cs.getConcept(3),"second obs value text");
+		//create a parent obs
+		Obs oParent =createObs(cs.getConcept(23), null); //in the concept set table as a set
+
+		//add o and o2 to the parent obs
+		oParent.addGroupMember(o2);
+		oParent.addGroupMember(o);
+
+		Encounter enc = buildEncounter();
+		enc.addObs(oParent);
+
+		enc = es.saveEncounter(enc);
+
+		//Now updating o
+		o.setValueText("Obs value updated");
+		enc = es.saveEncounter(enc);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		oParent = os.getObs(oParent.getObsId());
+		o = os.getObs(o.getObsId());
+		o2 = os.getObs(o2.getObsId());
+		Obs o3 = getNewVersionOfEditedObs(oParent,o);
+
+		assertEquals(1,obsAtTopLevelUpdated.size());
+		assertEquals(3, oParent.getGroupMembers(true).size());
+		assertTrue(o.getVoided());
+		assertFalse(oParent.getVoided());
+		assertFalse(o2.getVoided());
+
+		assertNotNull(o3);
+		assertFalse(o3.getVoided());
+		assertEquals("Obs value updated", o3.getValueText());
+	}
+
+	@Test
+	@Verifies(value = "should save encounter correctly when a level 1 child obs updated without recreating an entire obs hierarchy", method = "saveEncounter(Encounter)")
+	public void saveEncounter_shouldSaveEncounterWhenLevel1ObsIsUpdated(){
+		ConceptService cs = Context.getConceptService();
+		EncounterService es = Context.getEncounterService();
+		ObsService os = Context.getObsService();
+
+		Obs o1 = createObs(cs.getConcept(3),"original obs value text");
+		Obs o2 = createObs(cs.getConcept(3),"second obs value text");
+		Obs oParent =createObs(cs.getConcept(23), null); //in the concept set table as a set
+		Obs o3 = createObs(cs.getConcept(3), "third obs value text");
+
+		Obs oGrandParent = createObs(cs.getConcept(25), null);
+		oGrandParent.addGroupMember(oParent);
+		oGrandParent.addGroupMember(o3);
+
+		//add o and o2 to the parent obs
+		oParent.addGroupMember(o2);
+		oParent.addGroupMember(o1);
+
+		Encounter enc = buildEncounter();
+		enc.addObs(oGrandParent);
+
+		enc = es.saveEncounter(enc);
+
+		//Create a new obs o4 and add it to parent
+		Obs o4 = createObs(cs.getConcept(3), "fourth obs value text");
+		oParent = os.getObs(oParent.getObsId()); // Refresh the oParent
+
+		oParent.addGroupMember(o4);
+		enc = es.saveEncounter(enc);
+
+		Set<Obs> obsAtTopLevelUpdated = enc.getObsAtTopLevel(true);
+		assertEquals(1,obsAtTopLevelUpdated.size());
+
+		assertEquals(2, obsAtTopLevelUpdated.iterator().next().getGroupMembers(true).size()); // should be oParent and o3
+
+		oParent = os.getObs(oParent.getObsId());
+		assertFalse(oParent.getVoided());
+
+		o4 = os.getObs(o4.getObsId());
+		assertTrue(oParent.getGroupMembers(true).contains(o4));
+
```

Comment:
I don't see the assertions for the grand parent that they are not voided and still contain the same unvoided children except for the edited one